src/sources/kafka.rs

```diff
@@ -720,32 +1219,138 @@ fn create_consumer(config: &KafkaSourceConfig) -> crate::Result<StreamConsumer<C
     Ok(consumer)
 }
 
-#[derive(Default)]
-struct CustomContext {
+type TopicPartition = (String, i32);
+
+/// Status returned by partition consumer tasks, allowing the coordination task
+/// to differentiate between a consumer exiting normally (after receiving an end
+/// signal) and exiting when it reaches the end of a partition
+enum PartitionConsumerStatus {
+    NormalExit,
+    PartitionEOF,
+}
+
+enum KafkaCallback {
+    PartitionsAssigned(Vec<TopicPartition>, SyncSender<()>),
+    PartitionsRevoked(Vec<TopicPartition>, SyncSender<()>),
+    ShuttingDown(SyncSender<()>),
+}
+
+struct KafkaSourceContext {
+    acknowledgements: bool,
     stats: kafka::KafkaStatisticsContext,
-    finalizer: OnceLock<Arc<OrderedFinalizer<FinalizerEntry>>>,
+
+    /// A callback channel used to coordinate between the main consumer task and the acknowledgement task
+    callbacks: OnceLock<UnboundedSender<KafkaCallback>>,
```

Comment:
I see what's going on with `consumer` that needs a lock, but could this one be set up without the lock?